### PR TITLE
Gossipsub interop

### PR DESCRIPTION
**Pull request by dryajov**
rust interop fixes

### CODE REVIEW

File: libp2p/protocols/pubsub/gossipsub.nim [hunk 1]
```diff
@@ -199,25 +181,26 @@ proc getGossipPeers(g: GossipSub): Table[string, ControlMessage] {.gcsafe.} =
...
-    await sleepAsync(GossipSubHeartbeatInitialDelay)
+      for t in g.mesh.keys:
+        await g.rebalanceMesh(t)
 
-    for t in g.mesh.keys:
-      await g.rebalanceMesh(t)
+      await g.dropFanoutPeers()
+      let peers = g.getGossipPeers()
+      for peer in peers.keys:
+        await g.peers[peer].send(@[RPCMsg(control: some(peers[peer]))])
```
Comment (arnetheduck) [inline libp2p/protocols/pubsub/gossipsub.nim:195]: shouldn't there be a try/catch around here? a single peer failing should not cause the rest of the peers to not receive the message

File: libp2p/protocols/pubsub/gossipsub.nim [hunk 2]
```diff
@@ -413,39 +399,43 @@ method publish*(g: GossipSub,
...
+
+      let msg = newMessage(g.peerInfo, data, topic, g.sign)
+      var sent: seq[Future[void]]
+      for p in peers:
+        if p == g.peerInfo.id:
+          continue
+
+        trace "publishing on topic", name = topic
+        g.mcache.put(msg)
+        sent.add(g.peers[p].send(@[RPCMsg(messages: @[msg])]))
```
Comment (arnetheduck) [inline libp2p/protocols/pubsub/gossipsub.nim:425]: try/catch?

File: libp2p/protocols/pubsub/gossipsub.nim [hunk 3]
```diff
@@ -413,39 +399,43 @@ method publish*(g: GossipSub,
...
+        sent = await allFinished(sent)
```
Comment (arnetheduck) [inline libp2p/protocols/pubsub/gossipsub.nim:426]: shouldn't this be outside the loop?

File: libp2p/protocols/pubsub/gossipsub.nim (hunk 1 above)
Comment (dryajov) [reply libp2p/protocols/pubsub/gossipsub.nim:195]: actually, an `allFutures`/`allFinished` might be the right thing here?

Comment (dryajov) [reply libp2p/protocols/pubsub/gossipsub.nim:195]: but, yeah shouldn't stop broadcasting.

File: libp2p/protocols/pubsub/gossipsub.nim (hunk 2 above)
Comment (dryajov) [reply libp2p/protocols/pubsub/gossipsub.nim:425]: Tend wont throw right now and the futures are accumulated and then awaited with `allFutures` which won't throw either. Tho, I've reworked that part a bit.

File: libp2p/protocols/pubsub/gossipsub.nim (hunk 1 above)
Comment (sinkingsugar) [reply libp2p/protocols/pubsub/gossipsub.nim:195]: `allFinished` and warn if error I guess is the best here

File: libp2p/protocols/pubsub/gossipsub.nim (hunk 2 above)
Comment (sinkingsugar) [reply libp2p/protocols/pubsub/gossipsub.nim]: We removed `allFutures` from the codebase sometime ago, if possible stick with `allFinished` and the macro

File: libp2p/protocols/pubsub/gossipsub.nim [hunk 4]
```diff
@@ -442,10 +436,7 @@ method start*(g: GossipSub) {.async.} =
   ## start long running/repeating procedures
 
   # setup the heartbeat interval
-  g.heartbeatCancel = addInterval(GossipSubHeartbeatInterval,
-                                  proc (arg: pointer = nil)
-                                    {.gcsafe, locks: 0.} =
-                                    asyncCheck g.heartbeat)
+  g.heartbeatCancel = g.heartbeat()
```
Comment (sinkingsugar) [inline libp2p/protocols/pubsub/gossipsub.nim:439]: This change makes all much clearer 👍

File: libp2p/protocols/pubsub/gossipsub.nim (hunk 2 above)
Comment (dryajov) [reply libp2p/protocols/pubsub/gossipsub.nim]: its `allFinished`, typo.

**Pull request merged**